### PR TITLE
MIDI Jukebox

### PR DESCRIPTION
**Pull request by brndd**
## About the PR 

(Requires engine changes; I've not made the engine PR yet, but you can preview the changes here: https://github.com/space-wizards/RobustToolbox/compare/master...brndd:admin-midis)

Adds a MIDI jukebox that plays MIDI files stored on the server, streamed to the client with MIDI events, similar to instruments. The MIDIs are stored under the UserData directory on the server.

**Screenshots**

### External library

I'm using [DryWetMidi](https://github.com/melanchall/drywetmidi) for parsing and playing back MIDI files on the server. It is, as far as I can tell, by far the best native C# MIDI library, though unfortunately it does not have a software synthesizer (except, I believe, on Windows where it uses the Microsoft synthesizer).

The reason for going with an existing library should be obvious. Reinventing the wheel for something as complicated as MIDI would be a big waste of time, especially when there already exists robust native C# library to do it. Reimplementing the necessary features of this library would easily double or triple the amount of work required for this feature, and the result would in all likelihood be worse in quality.

### Events range & the stuck note issue

Currently the network events related to playing back MIDI files using this are broadcast to everyone. Ideally they would not be (to conserve bandwidth), but because MIDI is a sequential file format, moving in and out of range of the events *will* muck up playback with things like incorrect instruments and stuck notes. I've tried a few ways of solving this involving caching MIDI note events and program change events and sending those in Component State messages to no avail; the PVS system and other things related to this just aren't very friendly to this kind of thing currently.

There are two practical solutions I can think of:

- A new kind of event (or something) that lets the client know when it walks in and out of range of whatever filter the MIDI event sending is using (PVS, probably), so the client can stop all playing MIDI notes when this happens to prevent notes from getting stuck.
- Refactor of client-side MIDI code to use DryWetMidi (the external library being used server-side), which being a Better library is smart enough to know the length of NoteOn events when first sending them. This is *not* part of the MIDI spec, it is a convenience feature of the library; in the MIDI spec, you cannot know ahead of time how long a note is going to last, you only know how long it was when it's stopped by a Note Off message (and this fact is the root cause of this problem).
  - However, DryWetMidi doesn't have MIDI synthesizing capabilities except on Windows, so we can't use it to completely replace Fluidsynth. A possible solution would be to use DWM for sequencing and Fluidsynth for actually synthesizing.
  - DWM supports [custom output devices](https://melanchall.github.io/drywetmidi/articles/devices/Output-device.html#custom-output-device) which could probably be used to create an output device that wraps the FluidSynth synthesizer.
  - On that note it also supports custom input devices, but I have not looked into if it could be used for our needs.

### Buffering

I'd like to implement some buffering for the MIDI events to make playback more resilient to unstable network conditions. I've not yet figured out how best to implement this.

### Audio attenuation

Right now the range at which the juke is audible is... basically the entire station. I couldn't make heads or tails of our audio attenuation code and haven't gotten around to fixing this yet.

### Admin midis

As the branch name suggests, I originally wanted to reimplement the admin midis functionality from SS13 where admins could play arbitrary MIDIs globally for all players. I ended up making a jukebox instead, but some of the functionality here could perhaps be reused for admin MIDIs -- though admin MIDIs could possibly also just function more like instruments, where instead of streaming from server to client, the MIDI is streamed from the admin's client to the server and then to everyone else. This has the obvious problem of playback stopping if the admin goes offline.

**Changelog**

:cl:
- add: Added a MIDI jukebox for playing MIDI tunes in the bar.

### CODE REVIEW

File: Content.Client/Content.Client.csproj [hunk 1]
```diff
@@ -23,6 +23,11 @@
     <ProjectReference Include="..\RobustToolbox\Robust.Client\Robust.Client.csproj" />
     <ProjectReference Include="..\Content.Shared\Content.Shared.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Update="Juke\UI\MidiJukeWindow.xaml.cs">
+      <DependentUpon>MidiJukeWindow.xaml</DependentUpon>
+    </Compile>
+  </ItemGroup>
```
Comment (mirrorcult) [inline Content.Client/Content.Client.csproj:30]: can be removed

**Review by mirrorcult — COMMENTED** [review]
will do a more full review later

my main worry here is that the server->client midi streaming code is duplicated, since instruments presumably already do this (client->server->all clients in range).. wouldnt it be easier to pull that out into some intermediary thing so that instruments and midi juke can both use it? cause clearly instruments already have the attenuation/stop when out of range stuff figured out + we can maybe get rid of our dependency on drywetmidi, though i dont think its a big deal

**Comment by brndd** [comment]
> will do a more full review later

Cool, thanks. This is marked as a draft for a reason, so need to get into nitty gritty details unless you want to. There's still some work-in-progress code, such as a file in the engine part that's 3/4ths comment because I wasn't sure what I want to do with it yet.

> my main worry here is that the server->client midi streaming code is duplicated, since instruments presumably already do this (client->server->all clients in range).. wouldnt it be easier to pull that out into some intermediary thing so that instruments and midi juke can both use it? cause clearly instruments already have the attenuation/stop when out of range stuff figured out + we can maybe get rid of our dependency on drywetmidi, though i dont think its a big deal

Instrument code is... not really designed for this kind of modularity. Moving MIDI events to and forth over the network is the easy part -- it's everything else that's difficult. I did reuse what was possible from the existing MIDI system, which was mainly the MIDI renderer on the client. Another issue is that instruments are not ECS.

Instruments don't suffer from the stuck note problem because they use the non-ECS component messages thing, which sends the events to every client that has ever been in PVS range of the component. If instruments are ECS-ified and made to *not* send events globally, I'm quite certain they will suffer from the same stuck note issue.

**Comment by brndd** [comment]
As for the ramifications of sending MIDI events globally, it's mostly a minor bandwidth concern. Out of three random MIDI files I've looked at, their average bitrate (file size divided by length) is about 0.5 KiB/s, so that's approximately how much data is going over the pipe for each player. Double that to 1 KiB/s to make up for the fact that the MidiEvents we send over the network aren't as optimized as real MIDI, because the MIDI spec sometimes fits two fields into one byte while our events aren't as frugal.

MIDI compresses really well though so if bandwidth is ever a concern we can just compress the MIDI traffic. Zstd on a fast setting gets 80-90% file size reductions on MIDI files on my system.

**Comment by mirrorcult** [comment]
we already do compression for gamestate so compressing midi events is definitely a possibility

**Comment by brndd** [comment]
I should probably comment on this here as well, but doing some quick benchmarks with zstd it seems like compressing small amounts of MIDI data yields quite poor (1.1~1.4 compression ratio) results. I tested with a pregenerated dictionary based on a bunch of MIDI files I had laying around (`zstd -r --train -o mididict Music/midis`), and curiously results were *worse* with the pregenerated dictionary than with no dictionary at all -- so perhaps actual MIDI files on disk aren't good for training, and the training set should instead be based on individual MIDI events (as those are what's being sent over the network).

In any case, it is probably worth investigating this further; the zstd Github page has [a nice graph for a sample set of ~1KB data](https://github.com/facebook/zstd#the-case-for-small-data-compression) where a pretrained dictionary achieves a 9.921 compression ratio vs. a 2.8 compression ratio of no dictionary. Should probably try doing packet dumps of SS14 network traffic and train the dictionary on those to see if we gain compressibility that way.

**Comment by AJCM-git** [comment]
Re-open once you come back to it. Anybody else can pick this up if they want

**Pull request closed without merging**